group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2378,6 +2400,32 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
         return new CoordinatorResult<>(records, new StreamsGroupHeartbeatResult(response, internalTopicsToBeCreated));
     }
 
+    private void addToEndpointToPartitions(Set<Map.Entry<String, Set<Integer>>> taskEntrySet,
+                                           StreamsGroup group,
+                                           StreamsGroupHeartbeatResponseData.Endpoint responseEndpoint,
+                                           List<StreamsGroupHeartbeatResponseData.EndpointToPartitions> endpointToPartitionsList) {
+        for (Map.Entry<String, Set<Integer>> taskEntry : taskEntrySet) {
+            String subtopologyId = taskEntry.getKey();
+            List<Integer> partitions = new ArrayList<>(taskEntry.getValue());
+            ConfiguredSubtopology configuredSubtopology = group.configuredTopology().subtopologies().get(subtopologyId);
+            if (configuredSubtopology != null) {
+                List<StreamsGroupHeartbeatResponseData.TopicPartition> topicPartitions = Stream.concat(
```

Comment:
I think there is a corner case that needs to be handled. If you look at `PartitionGrouper` in the original `StreamsPartitionAssignor` code, there may be copartitioning groups with different number of partitions within the same subtopology. So while the number of tasks is always equal to the maximal number of partitions in all source topics, there maybe a subset of the source topics that have a lower number of partition. The example would be a subtopology that merges topic A with 1 partition and topic B with 2 partitions. We will get tasks 0 and 1, but we need to make sure that the topic partitions for task 0 is {A_0,B_0} and the topic partitions for task 1 is {B_1} (because topic A does not have a second partition).

To handle this corner case, it's probably enough to remove all elements of `partitions` that exceed the number of partitions in `topic`. 

I'd suggest we add a little class for this logic, so that we can test these corner cases independently of `GroupMetadatManager`.